### PR TITLE
[WIP] Add super-sample descriptor schema

### PR DESCRIPTION
**Pull request by mzganec**
## Checklist

* [ ] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [ ] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [ ] All inputs are used in process.
* [ ] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/descriptors/super_sample.yml [hunk 1]
```diff
@@ -0,0 +1,263 @@
+- slug: super_sample
```
Comment (mstajdohar) [inline resolwe_bio/descriptors/super_sample.yml]: This should be just `sample` or `sample2` or `sample-annotation-v2`.

File: resolwe_bio/descriptors/super_sample.yml [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  name: Super sample annotation
```
Comment (mstajdohar) [inline resolwe_bio/descriptors/super_sample.yml]: I'd call this Sample annotation.

File: resolwe_bio/descriptors/super_sample.yml [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  description: Sample meta data compliant with GEO and ENCODE databases.
+  version: 0.0.1
```
Comment (mstajdohar) [inline resolwe_bio/descriptors/super_sample.yml]: Let's make the first one 1.0.0, as it then looks more production ready to the end-users.

**Comment by mstajdohar** [comment]
When can we expect the reads schema?

File: resolwe_bio/descriptors/super_sample.yml [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+  schema:
+    - name: assay_type
+      label: Assay type
+      type: basic:string
+      choices:
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: If this is to be the new default descriptor schema for sample object, this list is not comprehensive enough. Please see this [example](https://github.com/genialis/resolwe-bio/blob/master/resolwe_bio/descriptors/reads.yml#L12) of the available choices in the GEO system. Explore also the latest and most up-to-date annotation $ submission template that can be dowloaded from GEO site. Use the `allow_custom_choice: true` for the choices menus where this is applicable.

File: resolwe_bio/descriptors/super_sample.yml [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        - label: RNA-seq
+          value: rna-seq
+        - label: ChIP-seq
+          value: chip-seq
+        - label: ATAC-seq
+          value: atac-seq
+        - label: ChIPmentation
+          value: chipmentation
+    - name: biosample_type
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: This choices list does not allow for proper classification of other possible types of samples, most notably microorganism cultures, but also other possible types that we might encounter down the road (e.g. metagenomic samples).

File: resolwe_bio/descriptors/super_sample.yml [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            - label: RNA binding protein
+              value: rna_binding_protein
+            - label: Chromatin remodeller
+              value: chromatin_remodeller
+            - label: Nucleotide modification
+              value: nucleotide_modification
+        - name: antibody_manufacturer
+          label: Antibody manufacturer
+          type: basic:string
+          choices:
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: This manufacturers list might be the most common, but its also too restrictive and without the `allow_custom_choice` option. There are also other manufacturers, some of them quite [big](https://www.sigmaaldrich.com/life-science/cell-biology/antibodies/antibody-products.html?TablePage=108208823)

File: resolwe_bio/descriptors/super_sample.yml [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+        - name: antibody_target
+          label: Antibody target
+          type: basic:string
+        - name: antibody_catalog_number
+          label: Antibody catalog number
+          type: basic:string
+        - name: antibody_lot_number
+          label: Antibody lot number
+          type: basic:string
+    - name: organism
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: Consider changing this to `species`. 

Too restrictive list without the `allow_custom_choice` option. The format of the `value` fields as envisioned will break everything that relies on the species name to be in binomial nomenclature. This affects gene knowlege-base, sample basket, existing data objects that will not match the sample basket restrictions (species field on expressions being `Homo sapiens` will not match `homo_sapiens` information on the samples.

File: resolwe_bio/descriptors/super_sample.yml [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+    - name: genotype
+      label: Genotype
+      type: basic:string
+    - name: biosample_treatment
+      label: Biosample treatment
+      type: basic:string
+    - name: treatment_protocol
+      label: Treatment protocol
+      type: basic:string
+    - name: platform
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: Too restrictive list without the `allow_custom_choice` option.

File: resolwe_bio/descriptors/super_sample.yml [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+          value: protein
+        - label: Other
+          value: other
+    - name: library_depleted_in
+      label: Library depleted in
+      type: basic:string
+    - name: library_treatment
+      label: Library treatment
+      type: basic:string
+    - name: replication_type
```
Comment (jkokosar) [inline resolwe_bio/descriptors/super_sample.yml]: What is this field for? In general, I think we lack `description` fields in this schema.

**Comment by jkokosar** [comment]
This schema does not allow for any custom (text) fields. Also, it lacks some commonly used meta-data fields like `gender` and `age`.  We should allow for some optional/general fields to cover cases where descriptor schema fields as envisioned here will turn out to not comprehensive enough.

If the new schema would include other [custom fields that are part of other descriptor schemas](https://github.com/genialis/resolwe-bio/blob/master/resolwe_bio/descriptors/sample_detailed.yml#L76), we could make those schemas obsolete and remove them entirely.

File: resolwe_bio/descriptors/super_sample.yml (hunk 4 above)
Comment (mzganec) [reply resolwe_bio/descriptors/super_sample.yml]: We added the option `Other` instead of custom user input.

**Comment by mzganec** [comment]
@jkokosar Please have a look.

File: resolwe_bio/descriptors/sample2.yml [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+        - name: library_treatment
+          label: Library treatment
+          type: basic:string
+          description: Describe how your library was treated.
+          required: false
+        - name: replication_type
+          label: Replication type
+          type: basic:string
+          description: >
+            Select what replication type you used, for more info, go to
```
Comment (jkokosar) [inline resolwe_bio/descriptors/sample2.yml]: Use hyperlinks

File: resolwe_bio/descriptors/sample2.yml [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+              value: tissue
+            - label: Primary cell
+              value: primary_cell
+            - label: Other
+              value: other
+        - name: primary_cell_line
+          label: Primary cell line
+          type: basic:string
+          description: >
+            Search for your tissue type: http://www.ontobee.org/ontology/UBERON and enter the UBERON
```
Comment (jkokosar) [inline resolwe_bio/descriptors/sample2.yml]: Use hyperlinks instead

File: resolwe_bio/descriptors/sample2.yml [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+          type: basic:string
+          description: >
+            Genotype of your sample, e.g. P53+/-, Tg2576, etc.
+          required: false
+        - name: biosample_type
+          label: Biosample type
+          type: basic:string
+          required: false
+          description: >
+            This is where you put what your samples are derived from, e.g. cell line, tissue,
```
Comment (jkokosar) [inline resolwe_bio/descriptors/sample2.yml]: This is where you put what your samples are derived.. Maybe it would be better "The source of samples/biological material, e.g. ..."

File: resolwe_bio/descriptors/sample2.yml (hunk 10 above)
Comment (mzganec) [reply resolwe_bio/descriptors/sample2.yml]: We don't support hyperlinks in description of inputs.
